Media & Text: Add media width control

Mamaduka opened the pull request:
## Description
Adds new control for media width. Using `RangeControl` with three preset options as suggested snap points. Props to @sarahmonster for the original idea.

Fixes #16853.
Fixes #17672.
Closes #14483. 

## How has this been tested?
1. Add the "Media & Text" block.
2. Set the media.
3. New "Media width" control should be visible in the "Media & Text settings" panel.
4. Changing the value in the new control should change the width of the media.

## Screenshots <!-- if applicable -->
<img width="281" alt="media-text-width-controls-no-marks" src="https://user-images.githubusercontent.com/240569/115536147-838b0c00-a2aa-11eb-8ad3-0b73e05ed32d.png">

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

Mamaduka committed:
Media & Text: Add media width control

paaljoachim commented:
Thank you George for creating the PR!

I brought up the PR during a Design triage.
We really like the new UI control.
I do believe it first needs to be added as a new general UI control. Then we can get it added to the Media & Text block.
@jasmussen Joen and @ItsJonQ Jon will likely have some input on this.

I believe this new UI control could also replace the similar control in the Image block.

jasmussen commented:
I've come to enjoy your PRs a lot, thank you for all your work 👌 

This is what I see:

![media text](https://user-images.githubusercontent.com/1204802/115522456-67787200-a28c-11eb-9db8-29a319ac6044.gif)

That's impressive work, it also looks cool and works reasonably well: the media and text scales only in percentages, so there isn't need for any unit selector.

There are a few things to consider, though; this is a new interface not seen anywhere else quite yet. It could potentially be leveraged for Image, which currently looks like this:

<img width="322" alt="Screenshot 2021-04-21 at 10 27 55" src="https://user-images.githubusercontent.com/1204802/115522738-aa3a4a00-a28c-11eb-8501-82c2ca70a981.png">

However, there's also a more consistent sidebar component design system outlined in #27331, and to summarize that up far more than it deserves to be, the goal is for sidebar controls to be far more generic across blocks — something you mostly toggle and it works, not something that you have to write software for, on a per-block basis. 

To put it even more simply, blocks should be able to say "I want dimension controls" (using a supports flag), and there they appear, looking the same across all blocks that opt in. #28356 goes into more detail on that. For that reason, I think the Media & Text block could ultimately a simpler control for dimensions, one that is not unique to it, but shares code and DNA with all other blocks that have sidebar scale controls (such as Spacer or Site Logo).

At the same time, this PR works well, isn't a lot of code, and solves a problem. So we might want to move forward with this PR, even if we also need to revisit it again in the future for further improvements.

However I don't think we should enable those scale-marks. I was honestly not even aware they were a part of the rangecontrol as it exists — but mostly because I haven't seen any blocks use that aspect yet, I think we should minimize the changes we do, and replicate what Spacer and Site Logo do: no scale marks. While less exciting, it makes it a smaller but still beneficial change, that doesn't preclude future iterations as part of #27331 and #28356. Make sense?

Thanks again!

Mamaduka committed:
Removes RangeControls marks

Mamaduka commented:
Thanks, @jasmussen, for the review and kind words ❤️

I've removed markers from the Range Control, and now the design should match Spacer and Site Logo size controls.

I'm all for generic and reusable controls for blocks. This new system will make things easier to maintain and reuse.

I've added issues you mentioned to my list and will go through them to have a future reference for similar changes.

P.S. I've updated the screenshot in the PR description.

jasmussen commented:
Honestly this one looks good to me! It could use a code sanity check, but 👍 👍 nice work!

paaljoachim commented:
I just tested the latest update. It looks nice and clean. I miss one thing though. 
There is a number but I do not know what it means. 

![Screen Shot 2021-04-21 at 20 49 22](https://user-images.githubusercontent.com/5323259/115605696-1e9fd800-a2e3-11eb-80c9-f1597305bfdd.png)

Should there be a percentage symbol next to the number?
(I also added the up/down arrow seen on hover to compare the space that is available.)

![Media-Text-block-Media-width](https://user-images.githubusercontent.com/5323259/115606251-c4ebdd80-a2e3-11eb-9712-08b4c2fa6811.jpg)

Should one be able to switch to other measurements like pixels, em or whatever else?
@aristath

Mamaduka commented:
Thanks, @paaljoachim, for testing the changes.

The `RangeControl` component doesn't have a unit option. The block also expects units for percentage, and change them to `px` or `rem` doesn't make sense.

paaljoachim commented:
@gwwar Kerry perhaps you could take a look at this PR.
Thanks!

gwwar reviewed (approved):
Thanks @Mamaduka code looks reasonable here and tests well. @jasmussen did we need a new issue/comment, or are the two noted issues https://github.com/WordPress/gutenberg/issues/27331 https://github.com/WordPress/gutenberg/issues/28356 enough to remind us about follow up work?

Mamaduka commented:
Thank you, @gwwar.
